src/invidious/connection/pool.cr

```diff
@@ -0,0 +1,90 @@
+module Invidious::ConnectionPool
+  struct Pool
+    property! url : URI
+    property! max_capacity : Int32
+    property! idle_capacity : Int32
+    property! timeout : Float64
+    property pool : DB::Pool(HTTP::Client)
+
+    def initialize(
+      url : URI,
+      *,
+      @max_capacity : Int32 = 5,
+      idle_capacity : Int32? = nil,
+      @timeout : Float64 = 5.0
+    )
+      if idle_capacity.nil?
+        @idle_capacity = @max_capacity
+      else
+        @idle_capacity = idle_capacity
+      end
+
+      @url = url
+
+      @pool = build_pool()
+    end
+
+    # Checks out a client in the pool
+    def client(&)
+      pool.checkout do |http_client|
+        # Proxy needs to be reinstated every time we get a client from the pool
+        http_client.proxy = make_configured_http_proxy_client() if CONFIG.http_proxy
+
+        response = yield http_client
+
+        return response
+      rescue ex : DB::Error
+        # Prevent broken client from being checked back into the pool
+        http_client.close
```

Comment:
https://github.com/crystal-lang/crystal-db/blob/892c25675634fcdf59780055c872be8468b2c204/src/db/pool.cr#L165

`HTTP::Client` lacks a `closed?` property so the pool won't actually remove it from the pool

We'll need to call pool.delete manually